### ISSUE TITLE
Promise never completed on RejectedExecutionException

### ISSUE THREAD

**Issue by scabug** [scala/bug#9071]
Rejected runnables from ExecutionContext do not complete Promises with Failures. If the underlying pool is overloaded and tasks are rejected (actually, if the rejection policy throws an exception) you get a Future that is never completed. Looking at the code I found that in CallbackRunnable.executeWithValue a NonFatal exception is catched and reported trough the configured reporter but nothing is done with the promise. I was expecting a Future completed with a Failure.

```scala
private class CallbackRunnable[T](val executor: ExecutionContext, val onComplete: Try[T] => Any) extends Runnable with OnCompleteRunnable {
  // must be filled in before running it
  var value: Try[T] = null

  override def run() = {
    require(value ne null) // must set value to non-null before running!
    try onComplete(value) catch { case NonFatal(e) => executor reportFailure e }
  }

  def executeWithValue(v: Try[T]): Unit = {
    require(value eq null) // can't complete it twice
    value = v
    // Note that we cannot prepare the ExecutionContext at this point, since we might
    // already be running on a different thread!
    try executor.execute(this) catch { case NonFatal(t) => executor reportFailure t }
  }
}
```

What I suggest is to pass the Promise to CallbackRunnable and complete it with a Failure if an exception is thrown when submitting the runnable. Doing that way the Failure will be propagated as usual. In order to do that, the onComplete method should be able to receive an optional Promise. I'll provide a patch.
Also, I'm not sure how useful would be to report a failure through reportFailure since the exception will be propagated as a Failure. I guess it can be removed too. What do you guys think about it?

**Comment by SethTisue** [comment]
the PR was abandoned. @viktorklang left some comments on Vladimir's efforts; perhaps someone else would like to pick this up? a first step would be to see if this is even still relevant after the `Future` changes in 2.12.

**Comment by NthPortal** [comment]
I believe @viktorklang has fixed this already - see [this commit](https://github.com/viktorklang/scala-futures/commit/2293c15bd9c55968bbb793868564fed9d98db7f1#diff-823e12da79dd0256861a4903168fab59R344)

**Comment by viktorklang** [comment]
@NthPortal @SethTisue Yes, I should really wrap my improvements up and propose them for 2.13

**Comment by viktorklang** [comment]
Thanks @gmethvin! Yes, this is closed and there is a regression test in scala-concurrent-tck

**Comment by coryfklein** [comment]
@viktorklang Do you have, or are you aware of plans to include this bug fix in a 2.12.x release?

**Comment by viktorklang** [comment]
@coryfklein It's not possible to backport the fix to 2.12.x since the 2.12.x implementation has strictly less information to work with. And I don't think (but haven't tried) it is viable to backport the entire new implementation to the 2.12.x series since I have a hard time seeing how it would be both forwards and backwards compatible with the previous 2.12 releases.
